### ISSUE TITLE
Are the schema validation checks within our webhooks redundant?

### ISSUE THREAD

**Issue by lyarwood** [kubevirt/kubevirt#12998]
/sig api

**What happened**:

This topic came up as part of the following thread:

https://groups.google.com/g/kubevirt-dev/c/4br7ZdejhI8

Using a bug to track further discussion and potentially handle the removal of this code if agreed on.

As of k8s v1.27 `kubectl` now defaults to enabling strict server side validation of object schema for each request:

https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2885-server-side-unknown-field-validation

Additionally with the introduction of `apiextensions.k8s.io/v1` pruning of unknown fields is additionally enabled for every CRD unless specifically disabled through the use of `x-kubernetes-preserve-unknown-fields: true`:

https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2332-pruning-for-custom-resources

These two factors combined essentially mean that by the time a request makes it to our webhooks the schema is already fully validated and thus make any further validation redundant:

https://github.com/kubevirt/kubevirt/blob/f99c707fc41b2ae30bc69c2959fd9e9ff1c2f8f9/pkg/util/openapi/openapi.go#L316-L331

The open question here is if we should drop this apparently redundant schema validation from our webhooks?

**Comment by xpivarc** [comment]
Reading through this...
So the server-side validation, including markers such as enums seems to be opt-in and clients are able to disable it? This is easily verified by running an old version of a client as that would not ask for the server-side validation, as far I see...

The thought always was: Keep our own schema validation because Kubernetes API is doing pruning and clients doing typos would just don't get what they ask for. The preserved unkown fields logic that @fossedihelm told me doesn't seem to allow us this, seems like we need to allow unkown fields and handle the invalidity by our self.

Note: Recently we introduced validation through marker that should be reverted...

**Comment by fossedihelm** [comment]
> The thought always was: Keep our own schema validation because Kubernetes API is doing pruning and clients doing typos would just don't get what they ask for. The preserved unkown fields logic that @fossedihelm told me doesn't seem to allow us this, seems like we need to allow unkown fields and handle the invalidity by our self.

If this is our goal, yes, we need to allow unknown fields and keep the current schema validation. Currently, the `+kubebuilder:pruning:PreserveUnknownFields` marker is only set in the objectmetadata of the CRD.

> 
> Note: Recently we introduced validation through marker that should be reverted...

Why should we revert them?

**Comment by lyarwood** [comment]
>> Reading through this... So the server-side validation, including markers such as enums seems to be opt-in and clients are able to disable it? 

Yes, FWIW I'm not sure that the enum validation is covered by our own schema check, definitely something to check.

>> This is easily verified by running an old version of a client as that would not ask for the server-side validation, as far I see...

Yes correct that's how I stumbled on this by accidentally using an older client with an unknown field and seeing that field pruned and the request accepted even with our existing schema validations in place.

>> The thought always was: Keep our own schema validation because Kubernetes API is doing pruning and clients doing typos would just don't get what they ask for. The preserved unkown fields logic that @fossedihelm told me doesn't seem to allow us this, seems like we need to allow unkown fields and handle the invalidity by our self.

While that makes logical sense to me it does go against the approach of k8s here with pruning of default fields now a default behaviour of api-server.

> If this is our goal, yes, we need to allow unknown fields and keep the current schema validation. Currently, the `+kubebuilder:pruning:PreserveUnknownFields` marker is only set in the objectmetadata of the CRD.
33

You can enable this in the spec of the overall CRD FWIW, the following PR was cleaning up our current but default setting of `false` for this field but we can instead enable this if agreed on:

https://github.com/kubevirt/kubevirt/pull/12991

>> Note: Recently we introduced validation through marker that should be reverted...

Link? I'm not entirely sure what you're referring to here sorry.

@EdDev @alaypatel07 - I've added this as a topic for next weeks unconference but if you can't make it we can try to bring it up in one of the sig-api calls that I always seem to have a conflict with.

https://unconference.kubevirt.io/D62SngbSRXO1uhTAc8Ikqg?both#lyarwood-sig-api-Enabling-PreserveUnknownFields

**Comment by xpivarc** [comment]
>Yes, FWIW I'm not sure that the enum validation is covered by our own schema check, definitely something to check.

I am not sure but we can easily try it out...
```
diff --git a/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go b/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go
index 2f989cec6d..1b9e1402fd 100644
--- a/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go
+++ b/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go
@@ -41,6 +41,7 @@ import (
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/runtime"
        "k8s.io/apimachinery/pkg/util/intstr"
+       "k8s.io/apimachinery/pkg/util/rand"
        "k8s.io/apimachinery/pkg/util/validation"
        k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
        "k8s.io/utils/ptr"
@@ -98,6 +99,43 @@ var _ = Describe("Validating VMICreate Admitter", func() {
                disableFeatureGates()
        })
 
+       It("should reject more than 256 disks", func() {
+               enableFeatureGate(virtconfig.HostDiskGate)
+               vmi := newBaseVmi()
+
+               for i := 0; i <= 256; i++ {
+                       name := rand.String(5)
+                       vmi.Spec.Domain.Devices.Disks = append(vmi.Spec.Domain.Devices.Disks, v1.Disk{
+                               Name: name,
+                       })
+                       vmi.Spec.Volumes = append(vmi.Spec.Volumes, v1.Volume{
+                               Name: name,
+                               VolumeSource: v1.VolumeSource{
+                                       HostDisk: &v1.HostDisk{
+                                               Path: "/root",
+                                               Type: v1.HostDiskExists,
+                                       },
+                               },
+                       })
+               }
+
+               vmiBytes, _ := json.Marshal(&vmi)
+
+               ar := &admissionv1.AdmissionReview{
+                       Request: &admissionv1.AdmissionRequest{
+                               Resource: webhooks.VirtualMachineInstanceGroupVersionResource,
+                               Object: runtime.RawExtension{
+                                       Raw: vmiBytes,
+                               },
+                       },
+               }
+
+               resp := vmiCreateAdmitter.Admit(context.Background(), ar)
+               Expect(resp.Allowed).To(BeFalse())
+       })
+
        It("should reject invalid VirtualMachineInstance spec on create", func() {
                vmi := newBaseVmi()
                vmi.Spec.Domain.Devices.Disks = append(vmi.Spec.Domain.Devices.Disks, v1.Disk{

```
It did not work for https://github.com/kubevirt/kubevirt/pull/11156 and I doubt it work for enum...

>While that makes logical sense to me it does go against the approach of k8s here with pruning of default fields now a default behaviour of api-server.

I believe the default was set just to preserve compatibility, same as for the server-side validation to allow clients to choose... but it doesn't really matter as we were strict in our API and now we are not.

**Comment by lyarwood** [comment]
/reopen

This was never resolved and still needs to be looked at IMHO.

Adding sig-control-plane that I'd still like to spin up for issues like this.

/sig control-plane
/remove-lifecycle rotten
